File: charts.py
"""
Chart generation module for all visualizations.
Handles: Creating Plotly charts with consistent theming
"""

import plotly.graph_objs as go
from config import COLORS


class ChartBuilder:
    """Builds Plotly charts with consistent styling"""
    
    def __init__(self, df, colors=None):
        """Initialize chart builder with data and colors"""
        self.df = df
        self.colors = colors or COLORS
    
    def create_stock_by_category_chart(self):
        """Create bar chart of stock quantity by category"""
        cat_data = self.df.groupby('Category')['Quantity'].sum().reset_index()
        fig = go.Figure(data=[
            go.Bar(x=cat_data['Category'], y=cat_data['Quantity'], marker_color=self.colors['primary'])
        ])
        fig.update_layout(
            title='Stock Quantity by Category',
            xaxis_title='Category',
            yaxis_title='Quantity',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig
    
    def create_stock_status_pie(self):
        """Create pie chart of stock status distribution"""
        status_data = self.df['Stock_Status'].value_counts()
        fig = go.Figure(data=[go.Pie(
            labels=status_data.index,
            values=status_data.values,
            hole=0.4,
            marker=dict(colors=[self.colors['danger'], self.colors['warning'], 
                               self.colors['success'], self.colors['primary']])
        )])
        fig.update_layout(
            title='Stock Status Distribution',
            paper_bgcolor=self.colors['card_bg'],
            plot_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig
    
    def create_value_by_warehouse_chart(self):
        """Create bar chart of total stock value by warehouse"""
        wh_data = self.df.groupby('Warehouse')['Total_Value'].sum().reset_index()
        fig = go.Figure(data=[
            go.Bar(x=wh_data['Warehouse'], y=wh_data['Total_Value'], marker_color=self.colors['success'])
        ])
        fig.update_layout(
            title='Total Stock Value by Warehouse',
            xaxis_title='Warehouse',
            yaxis_title='Value ($)',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig
    
    def create_top_products_chart(self):
        """Create bar chart of top 10 products by value"""
        top_products = self.df.nlargest(10, 'Total_Value')
        fig = go.Figure(data=[
            go.Bar(x=top_products['Product'], y=top_products['Total_Value'], 
                   marker_color=self.colors['warning'])
        ])
        fig.update_layout(
            title='Top 10 Products by Value',
            xaxis_title='Product',
            yaxis_title='Value ($)',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400,
            xaxis={'tickangle': -45}
        )
        return fig
    
    def create_carrying_cost_chart(self, carrying):
        """Create bar chart of carrying cost breakdown"""
        breakdown = carrying['breakdown']
        fig = go.Figure(data=[
            go.Bar(
                x=list(breakdown.keys()),
                y=list(breakdown.values()),
                marker_color=[self.colors['primary'], self.colors['info'], 
                             self.colors['warning'], self.colors['danger']],
                text=[f'${v:,.0f}' for v in breakdown.values()],
                textposition='auto'
            )
        ])
        fig.update_layout(
            title='Carrying Cost Components',
            xaxis_title='Cost Type',
            yaxis_title='Amount ($)',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig
    
    def create_supplier_performance_chart(self, supplier_perf):
        """Create dual-axis chart for supplier performance"""
        suppliers = list(supplier_perf['suppliers'].keys())
        scores = [supplier_perf['suppliers'][s]['quality_score'] for s in suppliers]
        values = [supplier_perf['suppliers'][s]['total_value'] for s in suppliers]
        
        fig = go.Figure()
        fig.add_trace(go.Bar(name='Quality Score', x=suppliers, y=scores, marker_color=self.colors['success']))
        fig.add_trace(go.Scatter(name='Total Value', x=suppliers, y=values, yaxis='y2', 
                                mode='lines+markers', marker=dict(size=10, color=self.colors['primary'])))
        
        fig.update_layout(
            title='Supplier Performance Analysis',
            xaxis_title='Supplier',
            yaxis_title='Quality Score',
            yaxis2=dict(title='Total Value ($)', overlaying='y', side='right'),
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400,
            legend=dict(x=0.01, y=0.99)
        )
        return fig
    
    def create_aging_chart(self, aging):
        """Create bar chart of item count by age"""
        fig = go.Figure(data=[
            go.Bar(
                x=list(aging['age_distribution'].keys()),
                y=list(aging['age_distribution'].values()),
                marker_color=[self.colors['success'], self.colors['info'], 
                             self.colors['warning'], self.colors['danger']],
                text=list(aging['age_distribution'].values()),
                textposition='auto'
            )
        ])
        fig.update_layout(
            title='Item Count by Age',
            xaxis_title='Age Range',
            yaxis_title='Number of Items',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig
    
    def create_aging_value_chart(self, aging):
        """Create bar chart of inventory value by age"""
        fig = go.Figure(data=[
            go.Bar(
                x=list(aging['value_by_age'].keys()),
                y=list(aging['value_by_age'].values()),
                marker_color=[self.colors['success'], self.colors['info'], 
                             self.colors['warning'], self.colors['danger']],
                text=[f'${v:,.0f}' for v in aging['value_by_age'].values()],
                textposition='auto'
            )
        ])
        fig.update_layout(
            title='Inventory Value by Age',
            xaxis_title='Age Range',
            yaxis_title='Value ($)',
            plot_bgcolor=self.colors['card_bg'],
            paper_bgcolor=self.colors['card_bg'],
            font={'color': self.colors['text']},
            height=400
        )
        return fig

File: static/js/charts.js
// Chart Rendering Module - All Plotly chart functions
// static/js/charts.js

const ChartRenderer = {
  colors: {
    primary: '#58a6ff',
    success: '#3fb950',
    warning: '#d29922',
    danger: '#f85149',
    muted: '#8b949e',
    background: '#161b22',
    text: '#c9d1d9'
  },

  defaultLayout: {
    paper_bgcolor: '#161b22',
    plot_bgcolor: '#161b22',
    font: { color: '#c9d1d9' },
    height: 400
  },

  // Stock Overview Charts
  renderStockStatus(elementId, stock) {
    const statusCounts = {};
    stock.forEach(item => {
      statusCounts[item.Stock_Status || 'Unknown'] = (statusCounts[item.Stock_Status] || 0) + 1;
    });

    const colorMap = {
      'Overstocked': this.colors.success,
      'Adequate': this.colors.primary,
      'Low': this.colors.warning,
      'Critical': this.colors.danger
    };

    Plotly.newPlot(elementId, [{
      values: Object.values(statusCounts),
      labels: Object.keys(statusCounts),
      type: 'pie',
      hole: 0.4,
      marker: { colors: Object.keys(statusCounts).map(s => colorMap[s] || this.colors.muted) }
    }], {
      ...this.defaultLayout,
      title: 'Stock Status Distribution'
    }, { responsive: true });
  },

  renderCategory(elementId, stock) {
    const catTotals = {};
    stock.forEach(item => {
      catTotals[item.Category || 'Unknown'] = (catTotals[item.Category] || 0) + (parseFloat(item.Quantity) || 0);
    });

    Plotly.newPlot(elementId, [{
      x: Object.keys(catTotals),
      y: Object.values(catTotals),
      type: 'bar',
      marker: { color: this.colors.primary }
    }], {
      ...this.defaultLayout,
      title: 'Stock Quantity by Category'
    }, { responsive: true });
  },

  renderWarehouse(elementId, stock) {
    const whTotals = {};
    stock.forEach(item => {
      whTotals[item.Warehouse || 'Unknown'] = (whTotals[item.Warehouse] || 0) + (parseFloat(item.Total_Value) || 0);
    });

    Plotly.newPlot(elementId, [{
      x: Object.keys(whTotals),
      y: Object.values(whTotals),
      type: 'bar',
      marker: { color: this.colors.success }
    }], {
      ...this.defaultLayout,
      title: 'Stock Value by Warehouse'
    }, { responsive: true });
  },

  renderTopProducts(elementId, stock) {
    const sorted = [...stock].sort((a, b) => (b.Total_Value || 0) - (a.Total_Value || 0)).slice(0, 10);

    Plotly.newPlot(elementId, [{
      y: sorted.map(p => p.Product),
      x: sorted.map(p => p.Total_Value),
      type: 'bar',
      orientation: 'h',
      marker: { color: this.colors.warning }
    }], {
      ...this.defaultLayout,
      title: 'Top 10 Products by Value',
      margin: { l: 150 }
    }, { responsive: true });
  },

  // KPI Charts
  renderCarryingCost(elementId, carrying) {
    if (!carrying?.breakdown) return;
    const breakdown = carrying.breakdown;

    Plotly.newPlot(elementId, [{
      x: Object.keys(breakdown),
      y: Object.values(breakdown),
      type: 'bar',
      marker: { color: [this.colors.primary, this.colors.success, this.colors.warning, this.colors.danger] },
      text: Object.values(breakdown).map(v => `$${v.toLocaleString()}`),
      textposition: 'auto'
    }], {
      ...this.defaultLayout,
      title: 'Carrying Cost Components'
    }, { responsive: true });
  },

  // Analytics Charts
  renderSupplierPerformance(elementId, supplierPerf) {
    if (!supplierPerf?.suppliers) {
      document.getElementById(elementId).innerHTML = '<div style="padding:40px;text-align:center;color:#8b949e">No supplier performance data available</div>';
      return;
    }

    const suppliers = Object.keys(supplierPerf.suppliers);
    const scores = suppliers.map(s => supplierPerf.suppliers[s].quality_score || 0);
    const values = suppliers.map(s => supplierPerf.suppliers[s].total_value || 0);

    Plotly.newPlot(elementId, [
      { name: 'Quality Score', x: suppliers, y: scores, type: 'bar', marker: { color: this.colors.success } },
      { name: 'Total Value', x: suppliers, y: values, type: 'scatter', mode: 'lines+markers', yaxis: 'y2', marker: { color: this.colors.primary, size: 10 } }
    ], {
      ...this.defaultLayout,
      title: 'Supplier Performance Analysis',
      yaxis: { title: 'Quality Score' },
      yaxis2: { title: 'Total Value ($)', overlaying: 'y', side: 'right' },
      legend: { x: 0.01, y: 0.99 }
    }, { responsive: true });
  },

  renderAging(countElementId, valueElementId, aging) {
    if (!aging) {
      document.getElementById(countElementId).innerHTML = '<div style="padding:40px;text-align:center;color:#8b949e">No aging data</div>';
      document.getElementById(valueElementId).innerHTML = '<div style="padding:40px;text-align:center;color:#8b949e">No aging data</div>';
      return;
    }

    const colors = [this.colors.success, this.colors.primary, this.colors.warning, this.colors.danger];

    if (aging.age_distribution) {
      Plotly.newPlot(countElementId, [{
        x: Object.keys(aging.age_distribution),
        y: Object.values(aging.age_distribution),
        type: 'bar',
        marker: { color: colors },
        text: Object.values(aging.age_distribution),
        textposition: 'auto'
      }], { ...this.defaultLayout, title: 'Item Count by Age' }, { responsive: true });
    }

    if (aging.value_by_age) {
      Plotly.newPlot(valueElementId, [{
        x: Object.keys(aging.value_by_age),
        y: Object.values(aging.value_by_age),
        type: 'bar',
        marker: { color: colors },
        text: Object.values(aging.value_by_age).map(v => `$${v.toLocaleString()}`),
        textposition: 'auto'
      }], { ...this.defaultLayout, title: 'Inventory Value by Age' }, { responsive: true });
    }
  },

  renderTrends(elementId, transactions) {
    const dates = {};
    transactions.forEach(t => {
      const date = t.Date?.split('T')[0] || t.Date?.split(' ')[0] || 'Unknown';
      dates[date] = (dates[date] || 0) + (parseFloat(t.Total_Value) || 0);
    });

    const sortedDates = Object.keys(dates).sort();
    Plotly.newPlot(elementId, [{
      x: sortedDates,
      y: sortedDates.map(d => dates[d]),
      type: 'scatter',
      mode: 'lines+markers',
      marker: { color: this.colors.primary },
      line: { color: this.colors.primary, width: 2 }
    }], {
      ...this.defaultLayout,
      title: 'Transaction Value Trends'
    }, { responsive: true });
  }
};

window.ChartRenderer = ChartRenderer;
